Revert "drivers: serial: nrfx_uarte: Deprecate CONFIG_UART_x_NRF_HW_A…

nordic-krch opened the pull request:
…SYNC"

This reverts commit 384e940c6665332d06aa939de15c7fc1479e79d7.

Feature was deprecated some time ago but we see that there is a need to keep it.

tomi-font commented on drivers/serial/Kconfig.nrfx_uart_instance:
```diff
@@ -49,23 +49,18 @@ config UART_$(nrfx_uart_num)_NRF_TX_BUFFER_SIZE
 	depends on HAS_HW_NRF_UARTE$(nrfx_uart_num)
 	depends on UART_ASYNC_API
 	depends on UART_NRFX_UARTE_LEGACY_SHIM
-	depends on !UART_NRFX_UARTE_ENHANCED_RX
```
Shouldn't this be kept?

nordic-krch replied:
I've split changes that are not part of revert into separate commits however the one pointed by you is part of the original commit.

tomi-font reviewed (changes requested):
https://github.com/zephyrproject-rtos/zephyr/pull/92156#discussion_r2177432550

nordic-krch committed:
Revert "drivers: serial: nrfx_uarte: Deprecate CONFIG_UART_x_NRF_HW_ASYNC"

This reverts commit 384e940c6665332d06aa939de15c7fc1479e79d7.

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
drivers: serial: nrfx_uarte: use ifdef instead of IS_ENABLED

In certain configurations timeout field is not present in the
data structure so preprocessor ifdef must be used instead of
compiler check.

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
drivers: serial: nrfx_uarte: Improve UART_NRFX_UARTE_ENHANCED_RX dependency

Enhanced RX cannot be used when UART_x_NRF_HW_ASYNC is used so changing
default value to depends on.

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

nordic-krch commented:
No. It's ok to pull it in after the release.

nordic-krch commented:
@kartben @danieldegrasse if it is not too late then that would be very good

danieldegrasse commented:
> @kartben @danieldegrasse if it is not too late then that would be very good

Is there a timeline to re-deprecate this feature? It would also be good to understand why we need to revive this- I guess the new method of byte counting still has some error conditions, so this method is needed?

dkalowsk commented:
@nordic-krch I'm okay with this going back in, but I'd like to ask for a release-note or migration guide update as well.  Specifically calling out why this is being re-enabled and the expected EOL for it.

nordic-krch commented:
We found out that new method has corner and it is not 100% reliable without HWFC and old approach is. So it's permanent de-deprecation. I will add the release note.

nordic-krch committed:
doc: releases: Add note about de-deprecation in Nordic UART driver

Add note in release notes for v4.2 about de-deprecation of
CONFIG_CONFIG_UART_x_NRF_HW_ASYNC option.

Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

dkalowsk reviewed (dismissed):
Awesome thanks for the update in the release-notes

danieldegrasse commented:
@nordic-krch can you rebase to remove the merge conflict

danieldegrasse commented:
@anangl can you revisit so we can get this merged for 4.2?
